pkg/operator/operator.go

```diff
@@ -743,3 +775,295 @@ func (o *Operator) Config(ctx context.Context, key string) (*manifests.Config, e
+	}
+
+	return true, nil
+}
+
+func (o *Operator) rebalanceWorkloads(ctx context.Context, namespace string, sel map[string]string) (bool, error) {
+	podList, err := o.client.ListPods(ctx, namespace, metav1.ListOptions{LabelSelector: labels.FormatLabels(sel)})
+	if err != nil {
+		return false, err
+	}
+
+	if len(podList.Items) <= 1 {
+		return false, nil
+	}
+
+	var (
+		pvcsToDelete []v1.PersistentVolumeClaim
+		podsToDelete []v1.Pod
+	)
+	for _, pod := range podList.Items {
+		for _, vol := range pod.Spec.Volumes {
+			if vol.PersistentVolumeClaim == nil {
+				continue
+			}
+			pvc, err := o.client.GetPersistentVolumeClaim(ctx, pod.Namespace, vol.PersistentVolumeClaim.ClaimName)
+			if apierrors.IsNotFound(err) {
+				break
+			}
+			if err != nil {
+				return false, err
+			}
+
+			dropPVC, ok := pvc.Annotations[dropPVCAnnotation]
+			if !ok {
+				break
+			}
+			if dropPVC == "yes" {
+				pvcsToDelete = append(pvcsToDelete, *pvc)
+				podsToDelete = append(podsToDelete, pod)
+			}
+		}
+	}
+
+	var workloadRebalanced int
+	for i, pod := range podsToDelete {
+		pvc := pvcsToDelete[i]
+		// Guard from deleting all PVCs to prevent complete data loss.
+		if workloadRebalanced == len(podList.Items)-1 {
+			// Remove the annotation so that the PVC doesn't get deleted in future cycles.
+			klog.V(4).Infof("Removing annotation %s from PersistentVolumeClaim %s/%s to avoid needlessly deleting all PVCs to rebalance a workload.", dropPVCAnnotation, pvc.Namespace, pvc.Name)
+			delete(pvc.Annotations, dropPVCAnnotation)
+			_, err := o.client.UpdatePersistentVolumeClaim(ctx, &pvc)
+			if err != nil {
+				return false, err
+			}
+			continue
+		}
+
+		klog.V(4).Infof("Rebalancing pod %s/%s.", pod.Namespace, pod.Name)
+		rebalanced, err := o.rebalanceWorkload(ctx, &pod, &pvc)
+		if err != nil {
+			return false, err
+		}
+
+		if rebalanced {
+			workloadRebalanced++
+		}
+	}
+
+	return workloadRebalanced > 0, nil
+}
+
+// workloadCorrectlySpread returns whether the selected pods are spread across
+// different nodes ensuring proper high-availability.
+// If the pods don't use persistent storage, it will always return true.
+func (o *Operator) workloadCorrectlySpread(ctx context.Context, namespace string, sel map[string]string) (bool, error) {
+	podList, err := o.client.ListPods(ctx, namespace, metav1.ListOptions{LabelSelector: labels.FormatLabels(sel)})
+	if err != nil {
+		return false, err
+	}
+
+	// Skip the check if we can't get enough pods. This prevents setting the status when the cluster is degraded.
+	if len(podList.Items) <= 1 {
+		return true, nil
+	}
+
+	// Skip workloads that don't have persistent storage enabled.
+	hasPVC := false
+	for _, vol := range podList.Items[0].Spec.Volumes {
+		if vol.PersistentVolumeClaim != nil {
+			hasPVC = true
+			break
+		}
+	}
+	if !hasPVC {
+		return true, nil
+	}
+
+	nodes := make(map[string]struct{}, len(podList.Items))
+	for _, pod := range podList.Items {
+		nodes[pod.Spec.NodeName] = struct{}{}
+	}
```

Comment:
`nodes` seems superfluous?
`return len(podList.Items) > 1, nil` achieves the same does it not?

Reply:
No, here we want to count the number of unique nodes hence why I am using a map of node names.